### PR TITLE
Check inputs num for node matching

### PR DESCRIPTION
**Pull request by xuzhenqi**
Fix inputs num mismatch for node matching. 

cc @justinchuby @gramalingam

### CODE REVIEW

**Comment by justinchuby** [comment]
Thanks for the PR. Could you provide a little bit of context on why this is needed?

File: onnxscript/rewriter/pattern.py [hunk 1]
```diff
@@ -954,6 +954,9 @@ def _match_node(self, pattern_node: NodePattern, node: ir.Node) -> bool:
 
         self._matched[pattern_node] = node
 
+        if len(node.inputs) != len(pattern_node.inputs):
```
Comment (gramalingam) [inline onnxscript/rewriter/pattern.py:957]: This is fine for now. We will need to revisit this to handle optional trailing inputs better.

**Review by gramalingam — APPROVED** [review]

File: onnxscript/rewriter/pattern.py (hunk 1 above)
Comment (justinchuby) [reply onnxscript/rewriter/pattern.py:957]: @xuzhenqi could you help add this as a comment in code?

**Comment by xuzhenqi** [comment]
> Thanks for the PR. Could you provide a little bit of context on why this is needed?

@justinchuby 

Suppose we need to remove an optional input of a node, for example, we want to replace `DFT(x, axis=-2)` with `DFT(x)`, since the default axis is -2, so we can remove the axis input. 

The solusion:

```python
class DFTSimplify(pat.RewriteRuleAsClass):
    @classmethod
    def pattern(
        cls, ope: pat.OpsetPatternBuilder, inp: pat.Var, axis: pat.Var, onesided: pat.Var
    ) -> pat.NodeOutputPattern:
        ret = ope.DFT(inp, None, axis, onesided=onesided)
        assert isinstance(ret, pat.NodeOutputPattern)
        return ret

    @classmethod
    def rewrite(
        cls, ope: pat.RewriterContext, inp: ir.Value, axis: ir.Value, onesided: ir.Attr
    ) -> ir.Value:
        del axis
        return ope.DFT(inp, dft_length=None, axis=None, onesided=onesided)

    @classmethod
    def check(cls, _context: None, inp: ir.Value, axis: ir.Value, onesided: ir.Attr) -> bool:
        del onesided
        logging.info("check")
        if axis.const_value is None:
            return False
        value = axis.const_value.numpy()
        axis_value = value.item()
        if axis_value < 0:
            return axis_value == -2
        if inp.shape is None:
            return False
        return axis_value + 2 == len(inp.shape)
```

The code will crash when matching a normal `DFT(x)` node, since there is no `axis` input to feed rewrite function.

**Comment by xuzhenqi** [comment]
@justinchuby  @gramalingam  Do I need to rebase the `main` branch?

**Comment by justinchuby** [comment]
Could you fix the lint errors? https://github.com/microsoft/onnxscript/actions/runs/11229454770/job/31245771279?pr=1885 Thanks

File: onnxscript/rewriter/pattern.py [hunk 2]
```diff
@@ -959,6 +959,10 @@
 
         self._matched[pattern_node] = node
 
+        # Note: Need to revisit this to handle optional trailing inputs better.
```
[the remark this answers is not in the document]
Comment (xuzhenqi) [reply onnxscript/rewriter/pattern.py]: fixed

[the remark this answers is not in the document]
Comment (xuzhenqi) [reply onnxscript/rewriter/pattern.py]: Fixed.

**Comment by xuzhenqi** [comment]
> Could you fix the lint errors? https://github.com/microsoft/onnxscript/actions/runs/11229454770/job/31245771279?pr=1885 Thanks

Fixed.

**Pull request merged**